core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -623,64 +750,12 @@ static void do_copy_state(struct crypto_hash_ctx *dst_ctx,
 
 		caam_cpy_block_src(&dst->blockbuf, &srcdata, 0);
 	}
-}
-
-/*
- * Registration of the hash Driver
- */
-static const struct crypto_hash_ops hash_ops = {
-	.init = do_init,
-	.update = do_update,
-	.final = do_final,
-	.free_ctx = do_free,
-	.copy_state = do_copy_state,
-};
 
-/*
- * Allocate the internal hashing data context
- *
- * @ctx    [out] Caller context variable
- * @algo   Algorithm ID
- */
-static TEE_Result caam_hash_allocate(struct crypto_hash_ctx **ctx,
-				     uint32_t algo)
-{
-	struct crypto_hash *hash = NULL;
-	struct hashdata *hashdata = NULL;
-	uint8_t hash_id = TEE_ALG_GET_MAIN_ALG(algo);
-	unsigned int algo_idx = hash_id - TEE_MAIN_ALGO_MD5;
-
-	HASH_TRACE("Allocate Context (%p) algo %" PRId16, ctx, algo_idx);
-
-	*ctx = NULL;
-
-	if (hash_id > caam_hash_limit)
-		return TEE_ERROR_NOT_IMPLEMENTED;
-
-	if (algo_idx > ARRAY_SIZE(hash_alg))
-		return TEE_ERROR_NOT_IMPLEMENTED;
-
-	hash = calloc(1, sizeof(*hash));
-	if (!hash)
-		return TEE_ERROR_OUT_OF_MEMORY;
-
-	hashdata = caam_calloc(sizeof(*hashdata));
-	if (!hashdata) {
-		HASH_TRACE("Allocation Hash data error");
-		free(hash);
-		return TEE_ERROR_OUT_OF_MEMORY;
+	if (src->key.length) {
+		if (caam_alloc_align_buf(&dst->key, dst->alg->size_key) ==
+		    CAAM_NO_ERROR)
+			memcpy(dst->key.data, src->key.data, src->key.length);
```

Comment:
**major:** If this allocation fails, how to handle the error? (as `copy_state()` does not return a status)

Reply:
Indeed, not safety net if the allocation fails.
We could ```panic()``` if it fails ?

Reply:
We already faced this: https://github.com/OP-TEE/optee_os/pull/3259#discussion_r322776210.

Reply:
Thanks for following up on this @etienne-lms.
This function originally called from `TEE_CopyOperation()` which cannot return an error in user space, so the TA will panic if we add an error code. But, it is much better to panic a TA than the entire TEE.
Perhaps there's something wrong in the design here (in the CAAM driver). The `dst->key.data` buffer could have been preallocated when the `struct hashctx` was allocated.

Reply:
Thanks @etienne-lms, I was not aware of this issue.
The ```dst``` hash context is also allocated in this function that would result in a ```panic()``` in case something goes wrong.
I though a ```panic()``` would be OK in ```caam_hash_hmac_copy_state()``` since it was merged like this in https://github.com/OP-TEE/optee_os/pull/3259

Reply:
I'd say panicking is safe as far as security is concerned but that may not be the smartest behavior on a OOM issue.

Reply:
On failure, maybe force `key.length = 0` to prevent the null buffer is used.
Or maybe each time `key.data` is about to be used, Core should report and error if `length != 0` but `data == NULL`.